src/ray/gcs/redis_context.cc

```diff
@@ -339,22 +335,22 @@ Status ConnectWithoutRetries(const std::string &address,
       oss << "Could not establish connection to Redis " << address << ":" << port
           << " (context.err = " << newContext->err << ").";
     }
-    return Status::RedisError(oss.str());
+    return std::make_pair(Status::RedisError(oss.str()), nullptr);
   }
-  RAY_CHECK(context) << "Unexpected null context pointer";
-  *context = newContext;
-  return Status::OK();
+  return std::make_pair(Status::OK(),
+                        std::unique_ptr<RedisContextType, RedisDeleter<RedisContextType>>(
+                            newContext, RedisDeleter<RedisContextType>()));
 }
 
-template <typename RedisContext, typename RedisConnectFunction>
-Status ConnectWithRetries(const std::string &address,
-                          int port,
-                          const RedisConnectFunction &connect_function,
-                          RedisContext **context) {
+template <typename RedisContextType, typename RedisConnectFunctionType>
+std::pair<Status, std::unique_ptr<RedisContextType, RedisDeleter<RedisContextType>>>
+ConnectWithRetries(const std::string &address,
+                   int port,
+                   const RedisConnectFunctionType &connect_function) {
   RAY_LOG(INFO) << "Attempting to connect to address " << address << ":" << port << ".";
   int connection_attempts = 0;
-  Status status = ConnectWithoutRetries(address, port, connect_function, context);
-  while (!status.ok()) {
+  auto resp = ConnectWithoutRetries<RedisContextType>(address, port, connect_function);
+  for (auto status = resp.first; !status.ok(); status = resp.first) {
```

Comment:
```suggestion
  auto [status, context] = ConnectWithoutRetries<RedisContextType>(address, port, connect_function);
  while (!status.ok()) {
```

I feel this is easier to read.

Reply:
oh nice

Reply:
Did it a little differently because by default this does copy, and one of the pair elements is `unique_ptr` (so can't be copied). `auto &[status, context]` works, but not sure how to do it with existing variables `[status, context]` inside the loop.

Reply:
Seems you already figured out a solution? 

```
auto [status, context] =
        std::move(ConnectWithRetries<redisContext>(ip_addresses[0], port, redisConnect));
```

Reply:
Yeah, that works here, but inside the loop it's not initialization, and it seems `[status, context] =` just doesn't work with existing variables?

Reply:
Also, I just saw an error in the build [here](https://buildkite.com/ray-project/oss-ci-build-pr/builds/44070#018c5729-bd23-49c3-8d9a-2b7e05464a9d) that says "moving a temporary prevents copy elision" so I think I will change it the old fashioned way on line 479 too.